grl/baselines/psr_jong.py

```diff
@@ -0,0 +1,323 @@
+        if len(indexes) == len(Q_check_m):
+            # all vectors are linearly independent, so we've found a new core test
+            Q.append(current)
+            Q_uvectors.append(current_u)
+
+            # add all neighbors to visit queue
+            for action in range(pomdp.n_actions):
+                for obs in range(pomdp.n_obs):
+                    new_pair = (action, obs)
+                    # append test consisting of [(a', o'), (a, o), ...] to to_visit
+                    to_visit.append([new_pair] + current)
+
+    # when we're done, return Q
+    return Q
+
+class PSR:
+    def __init__(self, pomdp, Q, pi):
+        self.num_Q = len(Q)
+        self.pi = pi
+        self.Q = Q
+        self.pred_vec = np.ones((self.num_Q,))
+
+        # set of extension tests and associated weight vectors
+        # currently zero-initalize? TODO
+        self.weights = {}
+        for action in range(pomdp.n_actions):
+            for obs in range(pomdp.n_obs): 
+                extension_pair = (action, obs)
+                for core_test in Q:
+                    ext_test = [extension_pair] + core_test
+                    self.weights[tuple(ext_test)] = np.zeros((self.num_Q,))
+
+    def get_pair_prob(self, action, observation):
+        """
+        Get the probability that the given observation followed the given action using the prediction vector.
+        Make sure you have updated the prediction vector according to the history first!
+        """
+        tup = ((action, observation),)
+        return np.transpose(self.pred_vec).dot(self.weights[tup])
+
+
+    def recalculate_prediction_vector(self, h):
+        """
+        Re-Calculates prediction vector p(Q|h) for a given history h.
+        A History is a list of (action, observation) pairs.
+        """
+        self.pred_vec = np.ones((self.num_Q,))
+
+        for step in h:
+            self.update_prediction_vector(step)
+
+
+        return self.pred_vec
+
+    def update_prediction_vector(self, pair):
+        """
+        Updates stored prediction vector with the given action-observation pair (tuple).
+        returns nothing.
+        """
+        base_test_weight = self.weights[(pair,)]
+        for i in range(self.num_Q):
+            ext_test_weight = self.weights[tuple([pair] + self.Q[i])]
+            self.pred_vec[i] = (np.transpose(self.pred_vec).dot(ext_test_weight)) / (np.transpose(self.pred_vec).dot(base_test_weight))
+            # bound as in original paper
+            self.pred_vec[i] = 1.0 if self.pred_vec[i] > 1.0 else self.pred_vec[i]
+            self.pred_vec[i] = 0.0001 if self.pred_vec[i] < 0.0001 else self.pred_vec[i]
+
+
+    def update_weights(self, history, stepsize):
+        # get all sub-histories using list comprehension, ignoring duplicates
+        # convert sub-histories to tuples for compatability with weight keys
+        sub_histories = set([tuple(history[i: j]) for i in range(len(history)) for j in range(i + 1, len(history) + 1)])
```

Comment:
If I had to guess, I think this is where the inefficiency is - I'm doing n^2 + n passes over the entire (currently up to 100k entries) history every weight update. Unsure how to improve this - I'm trying to look into how the original code improved this but it's hard to read.

Reply:
If there are many repeated histories you can get a big win by doing it from long to short: if some subsequence is in the sub_histories already, then so are all of _its_ subsequences.
```
sub_histories = set()
for i in range(len(history)):
    for j in reversed(range(i+1, len(history)+1)):
        subseq = tuple(history[i:j])
        if subseq in sub_histories:
            break
        sub_histories.add(subseq)
```

Reply:
It also seems as though you're re-computing this each time, when the previous set of sub-histories should comprise the majority of the current sub_histories. If that's the case, you really only need to look at sub_histories `[history[i:len(history)+1] for i in range(len(history))]`, and add those to the set. You can still use the logic above, starting from longest and breaking when you've seen one before. But I'm not sure how you want to handle the fact that histories are finite: once you have more than 100000, do you want to ignore histories that include observations you discarded? If so, then what I just said doesn't work.

Reply:
I think it would be fine to keep histories that were discarded; the 100k limit is just to save memory, the original paper doesn't mention throwing away past history (or even resetting the environment). I think keeping discarded sub-histories should be fine.

I'll try switching to your algorithm and storing computed sub-histories - I suspect that will make this much faster. Thanks for the help!

Reply:
(Side note: the original cpp also truncates the history; I assume that this was also to save memory as it wasn't mentioned anywhere in the paper.)